src/services/goals.js

```diff
@@ -2389,3 +2390,314 @@ export async function destroyGoal(goalIds) {
+        attributes: [
+          'id',
+          'goalId',
+        ],
+      },
+      {
+        model: GoalFieldResponse,
+        as: 'responses',
+        attributes: ['id', 'goalTemplateFieldPromptId', 'response'],
+      },
+      {
+        model: GoalResource,
+        as: 'goalResources',
+        attributes: ['id', 'resourceId'],
+      },
+      {
+        model: Objective,
+        as: 'objectives',
+        include: [
+          {
+            model: ObjectiveFile,
+            as: 'objectiveFiles',
+            attributes: ['id', 'fileId', 'objectiveId'],
+          },
+          {
+            model: ObjectiveResource,
+            as: 'objectiveResources',
+            attributes: ['id', 'resourceId', 'objectiveId'],
+          },
+          {
+            model: ObjectiveTopic,
+            as: 'objectiveTopics',
+            attributes: ['id', 'topicId', 'objectiveId'],
+          },
+          {
+            model: ActivityReportObjective,
+            as: 'activityReportObjectives',
+            attributes: ['id', 'objectiveId'],
+          },
+        ],
+      },
+    ],
+  });
+
+  const finalGoal = selectedGoals.find((goal) => goal.id === parseInt(finalGoalId, DECIMAL_BASE));
+
+  if (!finalGoal) {
+    throw new Error(`Goal with id ${finalGoalId} not found in merge goals`);
+  }
+
+  const finalGoalValues = determineFinalGoalValues(selectedGoals, finalGoal);
+
+  /**
+   * we will need to create a "new" final goal for each grant involved
+   * in this sordid business
+   */
+
+  const grantsWithReplacements = await Grant.findAll({
+    attributes: ['id', 'status', 'oldGrantId'],
+    where: {
+      [Op.or]: [
+        { id: uniq(selectedGoals.map((goal) => goal.grantId)) },
+        { oldGrantId: uniq(selectedGoals.map((goal) => goal.grantId)) },
+      ],
+      status: 'Active',
+    },
+  });
+
+  const grantsWithReplacementsDictionary = {};
+
+  grantsWithReplacements.forEach((grant) => {
+    grantsWithReplacementsDictionary[grant.oldGrantId || grant.id] = grant.id;
+  });
+
+  // unique list of grant IDs
+  const grantIds = uniq(grantsWithReplacements.map((grant) => grant.id));
+
+  const goalsToBulkCreate = grantIds.map((grantId) => ({
+    ...finalGoalValues,
+    grantId,
+  }));
+
+  const newGoals = await Goal.bulkCreate(goalsToBulkCreate, { individualHooks: true });
+
+  // we will need these in a moment
+  const grantToGoalDictionary = {};
+  newGoals.forEach((goal) => {
+    grantToGoalDictionary[goal.grantId] = goal.id;
+  });
+
+  // update associated models
+  // - update AR goals with originalGoalId and goalId with the new goalId
+  // - update goalId on objectives
+  // - copy existing objective, set mapsTo on existingObjective with new objectiveId
+  // - update ARO with originalObjectiveId and objectiveId with the new objectiveId
+  // - goalResources, copy
+  // - goalFieldResponses, copy
+  // - objective files, ETC
+
+  await Promise.all(selectedGoals.map((g) => (
```

Comment:
Maybe  comment on this one as it does lots of updating.

Reply:
sure